substrate/primitives/runtime/src/traits/transaction_extension/mod.rs

```diff
@@ -0,0 +1,590 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The transaction extension trait.
+
+use crate::{
+	scale_info::{MetaType, StaticTypeInfo},
+	transaction_validity::{TransactionValidity, TransactionValidityError, ValidTransaction},
+	DispatchResult,
+};
+use codec::{Codec, Decode, Encode};
+use impl_trait_for_tuples::impl_for_tuples;
+#[doc(hidden)]
+pub use sp_std::marker::PhantomData;
+use sp_std::{self, fmt::Debug, prelude::*};
+use sp_weights::Weight;
+use tuplex::{PopFront, PushBack};
+
+use super::{AccrueWeight, DispatchInfoOf, Dispatchable, OriginOf, PostDispatchInfoOf};
+
+mod as_transaction_extension;
+mod dispatch_transaction;
+#[allow(deprecated)]
+pub use as_transaction_extension::AsTransactionExtension;
+pub use dispatch_transaction::DispatchTransaction;
+
+/// Shortcut for the result value of the `validate` function.
+pub type ValidateResult<Val, Call> =
+	Result<(ValidTransaction, Val, OriginOf<Call>), TransactionValidityError>;
+
+/// Base for [TransactionExtension]s; this contains the associated types and does not require any
+/// generic parameterization.
+pub trait TransactionExtensionBase:
```

Comment:
What was again the reason for having two traits?

Reply:
So that the `implicit`, `weight`, `metadata` and identifier functionality is easily accessible without having to refer to the greater `TransactionExtension` trait, which needs the `Call` and `Context`, although I'm working on a definition to abstract away the `Context`. Most of these are used in proc macros where it's hard to disambiguate the type.

Reply:
Can you please given a concrete example of where this complicates things? A link to the code maybe?

Reply:
I gave a try in this PR https://github.com/paritytech/polkadot-sdk/pull/5440

I think one trait is also good.

in tests in `cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/lib.rs` we have to explicitly specify the runtime call type to get the implicit value of an extension whereas in the past it wasn't needed. But I think it is fine.

Reply:
It is not so relevant now that `fn weight` needs to take the call as parameter, and obviously the type must be known. This `weight` thing and the concrete call type actually created many problems, but now with the current architecture it's fixed, so we can merge the 2 traits, which has been done in #5440 .

Reply:
yes done in https://github.com/paritytech/polkadot-sdk/pull/3685/commits/56012d41452cd25d96260b8300a719ee322390f9